packages/contracts/src/token/ERC1155/RedeemOptimizerFIFOLIFO.sol

```diff
@@ -0,0 +1,299 @@
+            revert RedeemOptimizer__OptimizerFailed(i.amountFound, i.optimizerParams.amountToFind);
+        }
+
+        return _trimToSize(i.arrayIndex, i.depositPeriods, i.sharesAtPeriods);
+    }
+
+    /// @notice A struct to capture all the data that is iterated over. A means to reduce Stack Depth.
+    struct IterationData {
+        /// @dev The [IMultiTokenVault] we are querying against.
+        IMultiTokenVault vault;
+        /// @dev The [OptimizerParams] governing how we optimise for the redemption/withdrawal.
+        OptimizerParams optimizerParams;
+        /// @dev The current index of results written to the result array pair.
+        uint256 arrayIndex;
+        /// @dev The current Deposit Period which we are processing for.
+        uint256 depositPeriod;
+        /// @dev The sum of amounts found so far.
+        uint256 amountFound;
+        /// @dev The Deposit Period result array.
+        uint256[] depositPeriods;
+        /// @dev The Share Amounts At Period result array.
+        uint256[] sharesAtPeriods;
+        bool isDone;
+    }
+
+    /**
+     * @dev Encapsulates the processing of an iteration of one of the Period Ranges.
+     *
+     * @param i The [IterationData] capturing the current state of the processing.
+     * @return _updated The updated [IterationData].
+     */
+    function _iteration(IterationData memory i) internal view returns (IterationData memory _updated) {
+        uint256 sharesAtPeriod = i.vault.sharesAtPeriod(i.optimizerParams.owner, i.depositPeriod);
+        uint256 amountAtPeriod = i.optimizerParams.optimizerBasis == OptimizerBasis.Shares
+            ? sharesAtPeriod
+            : i.vault.convertToAssetsForDepositPeriod(sharesAtPeriod, i.depositPeriod, i.optimizerParams.redeemPeriod);
+
+        // If there is an Amount, store the value.
+        if (amountAtPeriod > 0) {
+            i.depositPeriods[i.arrayIndex] = i.depositPeriod;
+
+            // check if we will go "over" the Amount To Find.
+            if (i.amountFound + amountAtPeriod > i.optimizerParams.amountToFind) {
+                // we only need the amount that brings us to amountToFind
+                uint256 amountToInclude = i.optimizerParams.amountToFind - i.amountFound;
+
+                // in the assets case, the amounts include principal AND returns.  we want the shares on deposit,
+                // which is the principal only.
+                // use this ratio: partialShares / totalShares = partialAssets / totalAssets
+                //                 partialShares = (partialAssets * totalShares) / totalAssets
+                uint256 sharesToInclude = sharesAtPeriod.mulDiv(amountToInclude, amountAtPeriod);
+
+                // only include equivalent amount of shares for the amountToInclude assets
+                i.sharesAtPeriods[i.arrayIndex] =
+                    i.optimizerParams.optimizerBasis == OptimizerBasis.Shares ? amountToInclude : sharesToInclude;
+
+                i.isDone = true;
```

Comment:
@lucasia This is a valid exit criterion, but how do I update `i.amountFound` correctly here?